src/compiler/crystal/semantic/literal_expander.cr

```diff
@@ -28,32 +28,37 @@ module Crystal
     #     end
     def expand(node : ArrayLiteral)
       if node_of = node.of
-        if node.elements.size == 0
-          generic = Generic.new(Path.global("Array"), node_of).at(node)
-          call = Call.new(generic, "new").at(node)
-          return call
-        end
-
         type_var = node_of
       else
         type_var = TypeOf.new(node.elements.clone)
       end
 
       capacity = node.elements.size
 
-      buffer = new_temp_var.at(node)
+      generic = Generic.new(Path.global("Array"), type_var).at(node)
 
-      exps = Array(ASTNode).new(node.elements.size + 1)
-      node.elements.each_with_index do |elem, i|
-        exps << Call.new(buffer.clone, "[]=", NumberLiteral.new(i).at(node), elem.clone).at(node)
-      end
-      exps << NumberLiteral.new(capacity).at(node)
-      block_body = Expressions.new(exps).at(node)
+      if capacity.zero?
+        Call.new(generic, "new").at(node)
+      else
+        ary_var = new_temp_var.at(node)
 
-      block = Block.new([buffer.clone], block_body).at(node)
+        ary_instance = Call.new(generic, "unsafe_build", args: [NumberLiteral.new(capacity).at(node)] of ASTNode).at(node)
 
-      generic = Generic.new(Path.global("Array"), type_var).at(node)
-      Call.new(generic, "build", args: [NumberLiteral.new(capacity).at(node)] of ASTNode, block: block).at(node)
+        buffer = Call.new(ary_var, "to_unsafe")
+        buffer_var = new_temp_var.at(node)
+
+        exps = Array(ASTNode).new(node.elements.size + 3)
+        exps << Assign.new(ary_var, ary_instance).at(node)
```

Comment:
Shouldn't there be a clone for `ary_var` here? it's used in line 47 for a call also.

```suggestion
        exps << Assign.new(ary_var.clone, ary_instance).at(node)
```